substrate/frame/transaction-payment/skip-feeless-payment/src/lib.rs

```diff
@@ -104,67 +108,86 @@ impl<T, S> From<S> for SkipCheckIfFeeless<T, S> {
 	}
 }
 
-impl<T: Config + Send + Sync, S: SignedExtension<AccountId = T::AccountId>> SignedExtension
+pub enum Intermediate<T, O> {
+	/// The wrapped extension should be applied.
+	Apply(T),
+	/// The wrapped extension should be skipped.
+	Skip(O),
+}
+use Intermediate::*;
+
+impl<T: Config + Send + Sync, S: TransactionExtensionBase> TransactionExtensionBase
 	for SkipCheckIfFeeless<T, S>
-where
-	S::Call: CheckIfFeeless<Origin = frame_system::pallet_prelude::OriginFor<T>>,
 {
-	type AccountId = T::AccountId;
-	type Call = S::Call;
-	type AdditionalSigned = S::AdditionalSigned;
-	type Pre = (Self::AccountId, Option<<S as SignedExtension>::Pre>);
 	// From the outside this extension should be "invisible", because it just extends the wrapped
 	// extension with an extra check in `pre_dispatch` and `post_dispatch`. Thus, we should forward
 	// the identifier of the wrapped extension to let wallets see this extension as it would only be
 	// the wrapped extension itself.
 	const IDENTIFIER: &'static str = S::IDENTIFIER;
+	type Implicit = S::Implicit;
 
-	fn additional_signed(&self) -> Result<Self::AdditionalSigned, TransactionValidityError> {
-		self.0.additional_signed()
+	fn implicit(&self) -> Result<Self::Implicit, TransactionValidityError> {
+		self.0.implicit()
 	}
 
+	fn weight() -> frame_support::weights::Weight {
+		S::weight()
```

Comment:
This make the assumption that the weight for `call.is_feeless` is negligible.

Maybe the pallet macro should provide another attribute: `#[pallet::feeless_if::weight($expr)]` to allow user to give some specific weight.

Reply:
Yes this is a known limitation of this extension. I think it's fine to leave as is for now as custom origins set through `TransactionExtension`s will be the way to provide free transactions in the future anyway.